sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/PooledHashMap.java

```diff
@@ -0,0 +1,253 @@
+ * key is removed from the map.
+ *
+ * @param <K> The map key type
+ * @param <V> The map value type
+ */
+public class PooledHashMap<K, V> implements Map<K, V> {
+  private static final int DEFAULT_CAPACITY = 16;
+  private static final float LOAD_FACTOR = 0.75f;
+
+  private ArrayList<Entry<K, V>>[] table;
+  private final ObjectPool<Entry<K, V>> entryPool;
+  private int size;
+
+  /**
+   * Creates a {@link PooledHashMap} with {@code capacity} buckets.
+   *
+   * <p>The hashmap contains an array of buckets, each is an array-list of items. The number of
+   * buckets expands over time to avoid having too many items in one bucket, otherwise accessing an
+   * item by key won't be a constant time complexity.
+   *
+   * @param capacity The initial number of buckets to start with
+   */
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  public PooledHashMap(int capacity) {
+    this.table = new ArrayList[capacity];
+    this.entryPool = new ObjectPool<>(Entry::new);
+    this.size = 0;
+  }
+
+  /**
+   * Creates a new {@link PooledHashMap} with a default amount of buckets (capacity).
+   *
+   * @see PooledHashMap#PooledHashMap(int)
+   */
+  public PooledHashMap() {
+    this(DEFAULT_CAPACITY);
+  }
+
+  /**
+   * Add a key, value pair to the map.
+   *
+   * <p>Internally it uses a MapEntry from a pool of entries, to store this mapping
+   *
+   * @param key key with which the specified value is to be associated
+   * @param value value to be associated with the specified key
+   * @return Null if the was no previous mapping for this key, or the value of the previous mapping
+   *     of this key
+   */
+  @Override
+  @Nullable
+  public V put(K key, V value) {
+    Objects.requireNonNull(key, "This map does not support null keys");
+    Objects.requireNonNull(value, "This map does not support null values");
+    if (size > LOAD_FACTOR * table.length) {
+      rehash();
+    }
+
+    int bucket = getBucket(key);
+    ArrayList<Entry<K, V>> entries = table[bucket];
+    if (entries == null) {
+      entries = new ArrayList<>();
+      table[bucket] = entries;
+    } else {
+      for (int i = 0; i < entries.size(); i++) {
+        Entry<K, V> entry = entries.get(i);
+        if (Objects.equals(entry.key, key)) {
+          V oldValue = entry.value;
+          entry.value = value;
+          return oldValue;
+        }
+      }
+    }
+    Entry<K, V> entry = entryPool.borrowObject();
+    entry.key = key;
+    entry.value = value;
+    entries.add(entry);
+    size++;
+    return null;
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked", "NullAway"})
+  private void rehash() {
+    ArrayList<Entry<K, V>>[] oldTable = table;
+    table = new ArrayList[2 * oldTable.length];
+
+    // put() to new table below will reset size back to correct number
+    size = 0;
+
+    for (int i = 0; i < oldTable.length; i++) {
+      ArrayList<Entry<K, V>> bucket = oldTable[i];
+      if (bucket != null) {
+        bucket.forEach(
```

Comment:
I see a small reduction in allocations if you change this to a traditional for loop.

Reply:
Oh - I remember this. When I wrote this I was wonder under the impression buckets is not high cardinality, but after hitting the data structure book again I learned it was not the case to reach O(1). 
I'll fix.